tests/template_tests/syntax_tests/test_querystring.py

```diff
@@ -6,72 +6,79 @@
 
 
 class QueryStringTagTests(SimpleTestCase):
-    def setUp(self):
-        self.request_factory = RequestFactory()
-
-    @setup({"querystring_empty": "{% querystring %}"})
-    def test_querystring_empty(self):
-        request = self.request_factory.get("/")
-        template = self.engine.get_template("querystring_empty")
-        context = RequestContext(request)
+
+    request_factory = RequestFactory()
+
+    def assertRenderEqual(self, template_name, request=None, expected="", **context):
+        if request is None:
+            request = self.request_factory.get("/")
```

Comment:
I had to check why a couple of tests weren't using the helper
```suggestion
    def assertRenderEqual(self, template_name, request=None, context=None, expected=""):
        if context is None:
            context = {}
```
I think we can allow request as None as there are a few test cases surrounding this
Then we can use the helper consistently, with the last couple becoming like:
```python
    @setup({"querystring_query_dict": "{% querystring request.GET a=2 %}"})
    def test_querystring_with_explicit_query_dict(self):
        request = self.request_factory.get("/", {"a": 1})
        self.assertRenderEqual(
            "querystring_query_dict",
            context={"request": request},
            expected="?a=2",
        )

    @setup({"querystring_query_dict_no_request": "{% querystring my_query_dict a=2 %}"})
    def test_querystring_with_explicit_query_dict_and_no_request(self):
        context = {"my_query_dict": QueryDict("a=1&b=2")}
        self.assertRenderEqual(
            "querystring_query_dict_no_request",
            context=context,
            expected="?a=2&amp;b=2",
        )
```

Reply:
I'm not sure I follow this suggestion... The tests that I did not replace are using `render_to_string` instead of `Template.render` so I did not want to change that semantic. Specifically, I think that `test_querystring_with_explicit_query_dict_and_no_request` is intentionally not passing a request to the rendering, but if we migrate it to use the helper, it will start passing an instance of a `RequestContext` with a `request` being `None`, which is slightly different IMHO. To illustrate my point, suppose that the template tag code changes so instead of being this (current `main`):
```python
     if query_dict is None:
        query_dict = context.request.GET
```
it is changed to:
```python
    request_get = context.request.GET
     if query_dict is None:
        query_dict = request_get
```
Then the current `test_querystring_with_explicit_query_dict_and_no_request` fails with:
```
  File "~/fellowship/django/django/template/defaulttags.py", line 1195, in querystring
    request_get = context.request.GET
                  ^^^^^^^^^^^^^^^
AttributeError: 'Context' object has no attribute 'request'
```
But if we replace the `render_to_string` call, we get this instead:
```
  File "/home/nessita/fellowship/django/django/template/defaulttags.py", line 1195, in querystring
    request_get = context.request.GET
                  ^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'GET'
```

I would rather not change these two tests to ensure the original sentiment is kept, and also to not put that much logic in the assert helper (and risking loosing locality of behavior and readability of the tests).

Reply:
> > I would rather not change these two tests to ensure the original sentiment is kept
> 
> Ah I didn't notice the `render_to_string` distinction - makes sense. I think if this distinction is important, we should name the tests differently to emphasize this, otherwise it could be updated in future

I can certainly add a comment in each of these tests, in a third commit. Do you think that would suffice? Would you have some other suggestion?

> > also to not put that much logic in the assert helper (and risking loosing locality of behavior and readability of the tests).
> 
> It took me a while to understand what you meant here. I think what you're saying is we shouldn't update the helper to include a `render_to_string` case - I agree

Yes, correct! Sorry for not rephrasing it more clearly.

> As I had made a suggestion to the helper, it reads like you're against my suggestion because you believe it adds complexity and is less readable - which I disagree with

I did not mean that, and I agree with you that your suggestion does not add complexity. It's just that I tried to adapt the helper to use `Template.render` vs `render_to_string` depending on its arguments, and it started growing into something complex, so I reverted that attempt.

Reply:
Had a look into what `render_to_string` is doing and it's also a test helper shortcut which we can use in each test
So, `assertRenderEqual` can be:
```python
    def assertRenderEqual(self, template_name, context, expected):
        output = self.engine.render_to_string(template_name, context)
        self.assertEqual(output, expected)
```
and then tests look like:
```python
    @setup({"querystring_empty": "{% querystring %}"})
    def test_querystring_empty(self):
        request = self.request_factory.get("/")
        context = RequestContext(request)
        self.assertRenderEqual("querystring_empty", context, "")
```

Reply:
Thank you for looking into this. Following your comment, I also looked into `engine.render_to_string` and TIL that this is an independent method from `render_to_string`. But even with that in mind, I still think we can't easily unify the helper because I see a key difference in these tests: some are using a `RequestContext` (which is the case of current `assertRenderEqual`) and a couple are using a plain `Context` (which is the case of `engine.render_to_string`).

`RequestContext` is a subclass of `Context` that holds the request and its use also "guarantees" that context processors will populate it (contrary to what happens with plain `Context` variables). My original understanding of the tests is that the two that were using `engine.render_to_string` where intentionally mimicking the cases where `django.template.context_processors.request` is not enabled (as docs have an admonition about this).

So if my assumption is correct (@knyghty would you have some availability to confirm or deny?), I don't see a way to use the same helper for both "kind" of tests, since we want to distinguish between using a `Context` vs, using a `RequestContext` when rendering (and unifying this with an `if` is where the helper could get a tad overly complex and IMHO it defeats its purpose).

I hope the explanation of my understanding makes sense! :see_no_evil:

Reply:
`render_to_string` can use a `RequestContext` or a `Context` so there's no `if` in the helper?

Reply:
@sarahboyce I took a second look at this with a clearer head and your point made perfect sense. Thank you for the patience and guidance.

I have pushed these changes, let me know what you think! :rocket: :bow: